contracts/PushStaking/PushStaking.sol

```diff
@@ -0,0 +1,549 @@
+        }
+        uint256 sharesToBeAllocated = getSharesAmount(TotalShare, _percentage);
+        if (sharesToBeAllocated <= currentWalletShare) {
+            revert Errors.InvalidArg_LessThanExpected(currentWalletShare, sharesToBeAllocated);
+        }
+        walletShareInfo[_walletAddress].lastClaimedBlock = walletShareInfo[_walletAddress].lastClaimedBlock == 0
+            ? genesisEpoch
+            : walletShareInfo[_walletAddress].lastClaimedBlock;
+        _adjustWalletAndTotalStake(_walletAddress, sharesToBeAllocated, currentWalletShare);
+        WALLET_TOTAL_SHARES = TotalShare + sharesToBeAllocated;
+        emit NewSharesIssued(_walletAddress, sharesToBeAllocated);
+    }
+    /*
+     * Removing Wallet Share from a Wallet
+     * 1. removes the shares from a wallet completely
+     * 2. Can be called by governance.
+     * 3. Updates WALLET_TOTAL_SHARES
+     * 4. Emits out an event.
+    */
+
+    function removeWalletShare(address _walletAddress) public onlyGovernance {
+        if(_walletAddress == address(0) || _walletAddress == FOUNDATION) {
+            revert Errors.InvalidArgument_WrongAddress(_walletAddress);
+        }
+        uint256 sharesToBeRemoved = walletShareInfo[_walletAddress].walletShare;
+        _adjustWalletAndTotalStake(_walletAddress, 0, sharesToBeRemoved);
+        _adjustWalletAndTotalStake(FOUNDATION, sharesToBeRemoved, 0);
+
+        emit SharesRemoved(_walletAddress, sharesToBeRemoved);
+    }
+
+    function decreaseWalletShare(
+        address _walletAddress,
+        GenericTypes.Percentage memory _percentage
+    )
+        external
+        onlyGovernance
+    {
+        uint256 sharesToBeRemoved = walletShareInfo[_walletAddress].walletShare;
+        removeWalletShare(_walletAddress);
+        addWalletShare(_walletAddress, _percentage);
+        emit SharesDecreased(_walletAddress, sharesToBeRemoved, walletShareInfo[_walletAddress].walletShare);
+    }
+
+    /*
+     *Reward Calculation for a Given Wallet
+     * 1. calculateWalletRewards(address wallet)
+     * 2. public helper function
+     * 3. Helps in calculating rewards for a specific wallet based on
+       * a. Their Wallet Share 
+        * b. Total Wallet Shares
+        * c. Total Rewards available in WALLET_TOTAL_SHARES
+     * 4. Once calculated rewards for a sepcific wallet can be updated in WalletToRewards mapping
+     * 5. Reward can be calculated for wallets similar they are calculated for Token holders in a specific epoch.
+        * Reward for a Given Wallet X = ( Wallet Share of X / WALLET_TOTAL_SHARES) * WALLET_FEE_POOL
+    */
+    //TODO logic yet to be finalized
+    function calculateWalletRewards(address _wallet, uint256 _epochId) public view returns (uint256) {
+        return (walletShareInfo[_wallet].epochToWalletShares[_epochId] * epochRewardsForWallets[_epochId])
+            / epochToTotalShares[_epochId];
+    }
+
+    function claimShareRewards() external returns (uint256 rewards) {
```

Comment:
Critical To-Do:

- As previously discussed, let's ensure a wallet cannot claim or get rewards for the same epoch when they were added.
- Instead, if a Wallet is added at Epoch N, then the reward eligibility starts from Epoch N+1. @0xNilesh @Zartaj0